File: ui/lib/talents.ts
import fs from 'fs';
import path from 'path';

interface Dimensions {
  width: number;
  height: number;
}

export interface Talent {
  id: number;
  name: string;
  spellId: number;
  spellName: string;
  icon: string;
  ranks: Rank[];
}

export interface Rank {
  description: string;
  castTime?: string;
  powerCost?: string;
  range?: string;
  cooldown?: string;
}

export interface TalentNode {
  id: number;
  row: number;
  column: number;
  x: number;
  y: number;
  unlocks: number[];
  lockedBy: number[];
  maxRank: number;
  talents: Talent[];
  nodeType: string;
}

export interface TalentTree {
  className: string;
  classId: number;
  specName: string;
  specId: number;
  classSize: Dimensions;
  specSize: Dimensions;
  classNodes: TalentNode[];
  specNodes: TalentNode[];
  pvpTalents: Talent[];
}

function deserializeNode(jsonNode: any) {
  return {
    id: jsonNode['id'],
    row: jsonNode['row'],
    column: jsonNode['col'],
    x: jsonNode['x'],
    y: jsonNode['y'],
    unlocks: jsonNode['unlocks'],
    lockedBy: jsonNode['locked_by'],
    maxRank: jsonNode['max_rank'],
    talents: jsonNode['talents'].map(deserializeTalent),
    nodeType: jsonNode['node_type'],
  }
}

function deserializeTalent(jsonTalent: any) {
  return {
    id: jsonTalent['id'],
    name: jsonTalent['name'],
    spellId: jsonTalent['spell']['id'],
    spellName: jsonTalent['spell']['name'],
    icon: jsonTalent['icon'],
    ranks: jsonTalent['spell']['ranks'].map((jsonRank: any) => ({
      description: jsonRank['description'],
      castTime: jsonRank['cast_time'] ?? null,
      powerCost: jsonRank['power_cost'] ?? null,
      range: jsonRank['range'] ?? null,
      cooldown: jsonRank['cooldown'] ?? null,
    })),
  }
}

const wowDirectory = path.join(process.cwd(), 'wow');

export function getTalentTree(className: string, specName: string) {
  const fileName = `${className.toLowerCase()}-${specName.toLowerCase()}.json`.replace(' ', '-');
  const filePath = path.join(wowDirectory, 'talents', fileName);

  const fileContents = fs.readFileSync(filePath, 'utf8');
  const jsonTree = JSON.parse(fileContents);


  let classNodes = jsonTree['class_nodes'].map(deserializeNode);
  let specNodes = jsonTree['spec_nodes'].map(deserializeNode);


  const iconSize = 56;
  const iconPadding = 23;
  const paddedIconSize = iconSize + iconPadding * 2;
  const iconRows = 10;
  const iconCols = 9;

  // Tree width has an additional 0.5 columns to accommodate choice talents expanding.
  const width = paddedIconSize * (iconCols + 0.5);
  const height = paddedIconSize * iconRows;

  const tree: TalentTree = {
    className: jsonTree['class_name'],
    classId: jsonTree['class_id'],
    specName: jsonTree['spec_name'],
    specId: jsonTree['spec_id'],
    classNodes,
    specNodes,
    pvpTalents: jsonTree['pvp_talents'].map(deserializeTalent),
    classSize: {width, height},
    specSize: {width, height},
  }

  return tree;
}

File: pkg/serialize/talents.go
package serialize

import (
	"encoding/json"

	"github.com/crbednarz/moonkinmetrics/pkg/wow"
)

type talentTreeJson struct {
	ClassName  string           `json:"class_name"`
	SpecName   string           `json:"spec_name"`
	ClassNodes []talentNodeJson `json:"class_nodes"`
	SpecNodes  []talentNodeJson `json:"spec_nodes"`
	PvpTalents []talentJson     `json:"pvp_talents"`
	HeroTrees  []heroTreeJson   `json:"hero_trees"`
	ClassId    int              `json:"class_id"`
	SpecId     int              `json:"spec_id"`
}

type heroTreeJson struct {
	Name  string           `json:"name"`
	Nodes []talentNodeJson `json:"nodes"`
	Id    int              `json:"id"`
}

type talentNodeJson struct {
	Unlocks  []int        `json:"unlocks"`
	LockedBy []int        `json:"locked_by"`
	NodeType string       `json:"node_type"`
	Talents  []talentJson `json:"talents"`
	MaxRank  int          `json:"max_rank"`
	Id       int          `json:"id"`
	X        int          `json:"x"`
	Y        int          `json:"y"`
	Row      int          `json:"row"`
	Col      int          `json:"col"`
}

type talentJson struct {
	Name  string    `json:"name"`
	Icon  string    `json:"icon"`
	Spell spellJson `json:"spell"`
	Id    int       `json:"id"`
}

type spellJson struct {
	Name  string     `json:"name"`
	Ranks []rankJson `json:"ranks"`
	Id    int        `json:"id"`
}

type rankJson struct {
	Description string `json:"description"`
	CastTime    string `json:"cast_time"`
	PowerCost   string `json:"power_cost"`
	Range       string `json:"range"`
	Cooldown    string `json:"cooldown"`
}

func ExportTalentsToJson(talents *wow.TalentTree) ([]byte, error) {
	classNodes, err := convertNodes(talents.ClassNodes)
	if err != nil {
		return nil, err
	}

	specNodes, err := convertNodes(talents.SpecNodes)
	if err != nil {
		return nil, err
	}

	heroTrees := make([]heroTreeJson, len(talents.HeroTrees))
	for i, heroTree := range talents.HeroTrees {
		nodes, err := convertNodes(heroTree.Nodes)
		if err != nil {
			return nil, err
		}
		heroTrees[i] = heroTreeJson{
			Name:  heroTree.Name,
			Nodes: nodes,
			Id:    heroTree.Id,
		}
	}

	pvpTalents, err := convertTalents(talents.PvpTalents)
	if err != nil {
		return nil, err
	}

	tree := talentTreeJson{
		ClassName:  talents.ClassName,
		SpecName:   talents.SpecName,
		ClassId:    talents.ClassId,
		SpecId:     talents.SpecId,
		ClassNodes: classNodes,
		SpecNodes:  specNodes,
		PvpTalents: pvpTalents,
		HeroTrees:  heroTrees,
	}
	return json.MarshalIndent(tree, "", "  ")
}

func convertNodes(nodes []wow.TalentNode) ([]talentNodeJson, error) {
	jsonNodes := make([]talentNodeJson, len(nodes))
	for i, node := range nodes {
		talents, err := convertTalents(node.Talents)
		if err != nil {
			return nil, err
		}
		jsonNodes[i] = talentNodeJson{
			Id:       node.Id,
			X:        node.X,
			Y:        node.Y,
			Row:      node.Row,
			Col:      node.Col,
			LockedBy: node.LockedBy,
			Unlocks:  node.Unlocks,
			MaxRank:  node.MaxRank,
			NodeType: node.NodeType,
			Talents:  talents,
		}
	}
	return jsonNodes, nil
}

func convertTalents(talents []wow.Talent) ([]talentJson, error) {
	jsonTalents := make([]talentJson, len(talents))
	for i, talent := range talents {
		jsonTalents[i] = talentJson{
			Id:   talent.Id,
			Name: talent.Name,
			Icon: talent.Icon,
			Spell: spellJson{
				Id:    talent.Spell.Id,
				Name:  talent.Spell.Name,
				Ranks: convertRanks(talent.Spell.Ranks),
			},
		}
	}
	return jsonTalents, nil
}

func convertRanks(ranks []wow.Rank) []rankJson {
	jsonRanks := make([]rankJson, len(ranks))
	for i, rank := range ranks {
		jsonRanks[i] = rankJson{
			Description: rank.Description,
			CastTime:    rank.CastTime,
			PowerCost:   rank.PowerCost,
			Range:       rank.Range,
			Cooldown:    rank.Cooldown,
		}
	}
	return jsonRanks
}
